packages/legend-query-builder/src/stores/QueryBuilderChangeHistoryState.ts

```diff
@@ -0,0 +1,111 @@
+/**
+ * Copyright (c) 2020-present, Goldman Sachs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import type { RawLambda } from '@finos/legend-graph';
+import { ActionState, guaranteeNonNullable } from '@finos/legend-shared';
+import { makeObservable, observable, computed, action } from 'mobx';
+import type { QueryBuilderState } from './QueryBuilderState.js';
+
+export class QueryBuilderChangeHistoryState {
+  readonly queryBuilderState: QueryBuilderState;
+  readonly initState = ActionState.create();
+
+  querySnapshotBuffer: RawLambda[] = [];
+  currentQuery: RawLambda | undefined;
+  pointer = 0;
+  bufferSize = 10;
+
+  constructor(queryBuilderState: QueryBuilderState) {
+    makeObservable(this, {
+      currentQuery: observable,
+      querySnapshotBuffer: observable,
+      canUndo: computed,
+      canRedo: computed,
+      setCurrentQuery: action,
+      initialize: action,
+      undo: action,
+      redo: action,
+      pointer: observable,
+      cacheNewQuery: action,
```

Comment:
Do we have any style guide for organizing the values in a `makeObservable` call? I wonder if it would make sense to sort them alphabetically, or to group by `observable`, `computed`, and `action` types and sort alphabetically within those groups.